File: Kernel Simulator/Network/FTP/Filesystem/FTPFilesystem.vb

'    Kernel Simulator  Copyright (C) 2018-2021  EoflaOE
'
'    This file is part of Kernel Simulator
'
'    Kernel Simulator is free software: you can redistribute it and/or modify
'    it under the terms of the GNU General Public License as published by
'    the Free Software Foundation, either version 3 of the License, or
'    (at your option) any later version.
'
'    Kernel Simulator is distributed in the hope that it will be useful,
'    but WITHOUT ANY WARRANTY; without even the implied warranty of
'    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
'    GNU General Public License for more details.
'
'    You should have received a copy of the GNU General Public License
'    along with this program.  If not, see <https://www.gnu.org/licenses/>.

Imports System.IO
Imports System.Text

Module FTPFilesystem

    ''' <summary>
    ''' Lists remote folders and files
    ''' </summary>
    ''' <param name="Path">Path to folder</param>
    ''' <returns>The list if successful; null if unsuccessful</returns>
    ''' <exception cref="Exceptions.FTPFilesystemException"></exception>
    ''' <exception cref="InvalidOperationException"></exception>
    Public Function FTPListRemote(Path As String) As List(Of String)
        If connected Then
            Dim EntryBuilder As New StringBuilder
            Dim Entries As New List(Of String)
            Dim FileSize As Long
            Dim ModDate As Date
            Dim Listing As FtpListItem()

            Try
                If Path <> "" Then
                    Listing = ClientFTP.GetListing(Path)
                Else
                    Listing = ClientFTP.GetListing(currentremoteDir)
                End If
                For Each DirListFTP As FtpListItem In Listing
                    EntryBuilder.Append($"- {DirListFTP.Name}")
                    'Check to see if the file that we're dealing with is a symbolic link
                    If DirListFTP.Type = FtpObjectType.Link Then
                        EntryBuilder.Append(" >> ")
                        EntryBuilder.Append(DirListFTP.LinkTarget)
                        DirListFTP = DirListFTP.LinkObject
                    End If

                    If DirListFTP IsNot Nothing Then
                        If DirListFTP.Type = FtpObjectType.File Then
                            EntryBuilder.Append(": ")
                            FileSize = ClientFTP.GetFileSize(DirListFTP.FullName)
                            ModDate = ClientFTP.GetModifiedTime(DirListFTP.FullName)
                            EntryBuilder.Append(New Color(ListValueColor).VTSequenceForeground + DoTranslation("{0} KB | Modified in: {1}").FormatString(FormatNumber(FileSize / 1024, 2), ModDate.ToString))
                        ElseIf DirListFTP.Type = FtpObjectType.Directory Then
                            EntryBuilder.Append("/")
                        End If
                    End If
                    Entries.Add(EntryBuilder.ToString)
                    EntryBuilder.Clear()
                Next
                Return Entries
            Catch ex As Exception
                WStkTrc(ex)
                Throw New Exceptions.FTPFilesystemException(DoTranslation("Failed to list remote files: {0}"), ex, ex.Message)
            End Try
        Else
            Throw New InvalidOperationException(DoTranslation("You should connect to server before listing all remote files."))
        End If
        Return Nothing
    End Function

    ''' <summary>
    ''' Removes remote file or folder
    ''' </summary>
    ''' <param name="Target">Target folder or file</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    ''' <exception cref="Exceptions.FTPFilesystemException"></exception>
    Public Function FTPDeleteRemote(Target As String) As Boolean
        If connected Then
            Wdbg("I", "Deleting {0}...", Target)

            'Delete a file or folder
            If ClientFTP.FileExists(Target) Then
                Wdbg("I", "{0} is a file.", Target)
                ClientFTP.DeleteFile(Target)
            ElseIf ClientFTP.DirectoryExists(Target) Then
                Wdbg("I", "{0} is a folder.", Target)
                ClientFTP.DeleteDirectory(Target)
            Else
                Wdbg("E", "{0} is not found.", Target)
                Throw New Exceptions.FTPFilesystemException(DoTranslation("{0} is not found in the server."), Target)
                Return False
            End If
            Wdbg("I", "Deleted {0}", Target)
            Return True
        Else
            Throw New Exceptions.FTPFilesystemException(DoTranslation("You must connect to server with administrative privileges before performing the deletion."))
        End If
        Return False
    End Function

    ''' <summary>
    ''' Changes FTP remote directory
    ''' </summary>
    ''' <param name="Directory">Remote directory</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    ''' <exception cref="Exceptions.FTPFilesystemException"></exception>
    ''' <exception cref="InvalidOperationException"></exception>
    ''' <exception cref="ArgumentNullException"></exception>
    Public Function FTPChangeRemoteDir(Directory As String) As Boolean
        If connected = True Then
            If Directory <> "" Then
                If ClientFTP.DirectoryExists(Directory) Then
                    'Directory exists, go to the new directory
                    ClientFTP.SetWorkingDirectory(Directory)
                    currentremoteDir = ClientFTP.GetWorkingDirectory
                    Return True
                Else
                    'Directory doesn't exist, go to the old directory
                    Throw New Exceptions.FTPFilesystemException(DoTranslation("Directory {0} not found."), Directory)
                End If
            Else
                Throw New ArgumentNullException(Directory, DoTranslation("Enter a remote directory. "".."" to go back"))
            End If
        Else
            Throw New InvalidOperationException(DoTranslation("You must connect to a server before changing directory"))
        End If
        Return False
    End Function

    Public Function FTPChangeLocalDir(Directory As String) As Boolean
        If Directory <> "" Then
            Dim targetDir As String
            targetDir = $"{currDirect}/{Directory}"

#If NTFSCorruptionFix Then
            'Mitigate Windows 10 NTFS corruption or Windows 10 BSOD bug
            If IsOnWindows() And (targetDir.Contains("$i30") Or targetDir.Contains("\\.\globalroot\device\condrv\kernelconnect")) Then
                Wdbg("F", "Trying to access invalid path. Path was {0}", targetDir)
                Throw New ArgumentException(DoTranslation("Trying to access invalid path."))
            End If
#End If

            'Check if folder exists
            If IO.Directory.Exists(targetDir) Then
                'Parse written directory
                Dim parser As New IO.DirectoryInfo(targetDir)
                currDirect = parser.FullName
                Return True
            Else
                Throw New Exceptions.FTPFilesystemException(DoTranslation("Local directory {0} doesn't exist."), Directory)
            End If
        Else
            Throw New ArgumentNullException(Directory, DoTranslation("Enter a local directory. "".."" to go back."))
        End If
        Return False
    End Function

    ''' <summary>
    ''' Move file or directory to another area, or rename the file
    ''' </summary>
    ''' <param name="Source">Source file or folder</param>
    ''' <param name="Target">Target file or folder</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    ''' <exception cref="InvalidOperationException"></exception>
    Public Function FTPMoveItem(Source As String, Target As String) As Boolean
        If connected Then
            Dim Success As Boolean

            'Begin the moving process
            Dim SourceFile As String = Source.Split("/").Last
            Wdbg("I", "Moving from {0} to {1} with the source file of {2}...", Source, Target, SourceFile)
            If ClientFTP.DirectoryExists(Source) Then
                Success = ClientFTP.MoveDirectory(Source, Target)
            ElseIf ClientFTP.FileExists(Source) And ClientFTP.DirectoryExists(Target) Then
                Success = ClientFTP.MoveFile(Source, Target + SourceFile)
            ElseIf ClientFTP.FileExists(Source) Then
                Success = ClientFTP.MoveFile(Source, Target)
            End If
            Wdbg("I", "Moved. Result: {0}", Success)
            Return Success
        Else
            Throw New InvalidOperationException(DoTranslation("You must connect to server before performing transmission."))
        End If
        Return False
    End Function

    ''' <summary>
    ''' Copy file or directory to another area, or rename the file
    ''' </summary>
    ''' <param name="Source">Source file or folder</param>
    ''' <param name="Target">Target file or folder</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    ''' <exception cref="InvalidOperationException"></exception>
    Public Function FTPCopyItem(Source As String, Target As String) As Boolean
        If connected Then
            Dim Success As Boolean = True
            Dim Result As Object

            'Begin the copying process
            'TODO: FluentFTP currently doesn't support .CopyFile and .CopyDirectory
            Dim SourceFile As String = Source.Split("/").Last
            Wdbg("I", "Copying from {0} to {1} with the source file of {2}...", Source, Target, SourceFile)
            If ClientFTP.DirectoryExists(Source) Then
                ClientFTP.DownloadDirectory(paths("Temp") + "/FTPTransfer", Source)
                Result = ClientFTP.UploadDirectory(paths("Temp") + "/FTPTransfer/" + Source, Target)
            ElseIf ClientFTP.FileExists(Source) And ClientFTP.DirectoryExists(Target) Then
                ClientFTP.DownloadFile(paths("Temp") + "/FTPTransfer/" + SourceFile, Source)
                Result = ClientFTP.UploadFile(paths("Temp") + "/FTPTransfer/" + SourceFile, Target + "/" + SourceFile)
            ElseIf ClientFTP.FileExists(Source) Then
                ClientFTP.DownloadFile(paths("Temp") + "/FTPTransfer/" + SourceFile, Source)
                Result = ClientFTP.UploadFile(paths("Temp") + "/FTPTransfer/" + SourceFile, Target)
            End If
            Directory.Delete(paths("Temp") + "/FTPTransfer", True)

            'See if copied successfully
#Disable Warning BC42104
            If Result.GetType = GetType(List(Of FtpResult)) Then
                For Each FileResult As FtpResult In Result
                    If FileResult.IsFailed Then
                        Wdbg("E", "Transfer for {0} failed: {1}", FileResult.Name, FileResult.Exception.Message)
                        WStkTrc(FileResult.Exception)
                        Success = False
                    End If
                Next
            ElseIf Result.GetType = GetType(FtpStatus) Then
                If CType(Result, FtpStatus).IsFailure Then
                    Wdbg("E", "Transfer failed")
                    Success = False
                End If
            End If
#Enable Warning BC42104
            Wdbg("I", "Copied. Result: {0}", Success)
            Return Success
        Else
            Throw New InvalidOperationException(DoTranslation("You must connect to server before performing transmission."))
        End If
        Return False
    End Function

    ''' <summary>
    ''' Changes the permissions of a remote file
    ''' </summary>
    ''' <param name="Target">Target file</param>
    ''' <param name="Chmod">Permissions in CHMOD format. See https://man7.org/linux/man-pages/man2/chmod.2.html chmod(2) for more info.</param>
    ''' <returns>True if successful; False if unsuccessful</returns>
    Public Function FTPChangePermissions(Target As String, Chmod As Integer) As Boolean
        If connected Then
            Try
                ClientFTP.Chmod(Target, Chmod)
                Return True
            Catch ex As Exception
                Wdbg("E", "Error setting permissions ({0}) to file {1}: {2}", Chmod, Target, ex.Message)
                WStkTrc(ex)
            End Try
        Else
            Throw New InvalidOperationException(DoTranslation("You must connect to server before performing filesystem operations."))
        End If
        Return False
    End Function

End Module

File: Kernel Simulator/Network/FTP/Filesystem/FTPFilesystem.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using FluentFTP;
using FluentFTP.Helpers;
using KS.ConsoleBase.Colors;
using KS.Files;

// Kernel Simulator  Copyright (C) 2018-2022  Aptivi
// 
// This file is part of Kernel Simulator
// 
// Kernel Simulator is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// Kernel Simulator is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

using KS.Files.Querying;
using KS.Languages;
using KS.Misc.Text;
using KS.Misc.Writers.DebugWriters;

namespace KS.Network.FTP.Filesystem
{
	public static class FTPFilesystem
	{

		/// <summary>
		/// Lists remote folders and files
		/// </summary>
		/// <param name="Path">Path to folder</param>
		/// <returns>The list if successful; null if unsuccessful</returns>
		/// <exception cref="Exceptions.FTPFilesystemException"></exception>
		/// <exception cref="InvalidOperationException"></exception>
		public static List<string> FTPListRemote(string Path)
		{
			return FTPListRemote(Path, FTPShellCommon.FtpShowDetailsInList);
		}

		/// <summary>
		/// Lists remote folders and files
		/// </summary>
		/// <param name="Path">Path to folder</param>
		/// <param name="ShowDetails">Shows the details of the file</param>
		/// <returns>The list if successful; null if unsuccessful</returns>
		/// <exception cref="Exceptions.FTPFilesystemException"></exception>
		/// <exception cref="InvalidOperationException"></exception>
		public static List<string> FTPListRemote(string Path, bool ShowDetails)
		{
			if (FTPShellCommon.FtpConnected)
			{
				var EntryBuilder = new StringBuilder();
				var Entries = new List<string>();
				long FileSize;
				DateTime ModDate;
				FtpListItem[] Listing;

				try
				{
					if (!string.IsNullOrEmpty(Path))
					{
						Listing = FTPShellCommon.ClientFTP.GetListing(Path);
					}
					else
					{
						Listing = FTPShellCommon.ClientFTP.GetListing(FTPShellCommon.FtpCurrentRemoteDir);
					}
					foreach (FtpListItem DirListFTP in Listing)
					{
						var listItem = DirListFTP;
						EntryBuilder.Append($"- {listItem.Name}");
						// Check to see if the file that we're dealing with is a symbolic link
						if (listItem.Type == FtpObjectType.Link)
						{
							EntryBuilder.Append(" >> ");
							EntryBuilder.Append(listItem.LinkTarget);
							listItem = listItem.LinkObject;
						}

						if (listItem is not null)
						{
							if (listItem.Type == FtpObjectType.File)
							{
								if (ShowDetails)
								{
									EntryBuilder.Append(": ");
									FileSize = FTPShellCommon.ClientFTP.GetFileSize(listItem.FullName);
									ModDate = FTPShellCommon.ClientFTP.GetModifiedTime(listItem.FullName);
									EntryBuilder.Append(KernelColorTools.ListValueColor.VTSequenceForeground + Translate.DoTranslation("{0} KB | Modified in: {1}").FormatString((FileSize / 1024d).ToString("N2"), ModDate.ToString()));
								}
							}
							else if (listItem.Type == FtpObjectType.Directory)
							{
								EntryBuilder.Append("/");
							}
						}
						Entries.Add(EntryBuilder.ToString());
						EntryBuilder.Clear();
					}
					return Entries;
				}
				catch (Exception ex)
				{
					DebugWriter.WStkTrc(ex);
					throw new Kernel.Exceptions.FTPFilesystemException(Translate.DoTranslation("Failed to list remote files: {0}"), ex, ex.Message);
				}
			}
			else
			{
				throw new InvalidOperationException(Translate.DoTranslation("You should connect to server before listing all remote files."));
			}
			return null;
		}

		/// <summary>
		/// Removes remote file or folder
		/// </summary>
		/// <param name="Target">Target folder or file</param>
		/// <returns>True if successful; False if unsuccessful</returns>
		/// <exception cref="Exceptions.FTPFilesystemException"></exception>
		public static bool FTPDeleteRemote(string Target)
		{
			if (FTPShellCommon.FtpConnected)
			{
				DebugWriter.Wdbg(DebugLevel.I, "Deleting {0}...", Target);

				// Delete a file or folder
				if (FTPShellCommon.ClientFTP.FileExists(Target))
				{
					DebugWriter.Wdbg(DebugLevel.I, "{0} is a file.", Target);
					FTPShellCommon.ClientFTP.DeleteFile(Target);
				}
				else if (FTPShellCommon.ClientFTP.DirectoryExists(Target))
				{
					DebugWriter.Wdbg(DebugLevel.I, "{0} is a folder.", Target);
					FTPShellCommon.ClientFTP.DeleteDirectory(Target);
				}
				else
				{
					DebugWriter.Wdbg(DebugLevel.E, "{0} is not found.", Target);
					throw new Kernel.Exceptions.FTPFilesystemException(Translate.DoTranslation("{0} is not found in the server."), Target);
					return false;
				}
				DebugWriter.Wdbg(DebugLevel.I, "Deleted {0}", Target);
				return true;
			}
			else
			{
				throw new Kernel.Exceptions.FTPFilesystemException(Translate.DoTranslation("You must connect to server with administrative privileges before performing the deletion."));
			}
			return false;
		}

		/// <summary>
		/// Changes FTP remote directory
		/// </summary>
		/// <param name="Directory">Remote directory</param>
		/// <returns>True if successful; False if unsuccessful</returns>
		/// <exception cref="Exceptions.FTPFilesystemException"></exception>
		/// <exception cref="InvalidOperationException"></exception>
		/// <exception cref="ArgumentNullException"></exception>
		public static bool FTPChangeRemoteDir(string Directory)
		{
			if (FTPShellCommon.FtpConnected == true)
			{
				if (!string.IsNullOrEmpty(Directory))
				{
					if (FTPShellCommon.ClientFTP.DirectoryExists(Directory))
					{
						// Directory exists, go to the new directory
						FTPShellCommon.ClientFTP.SetWorkingDirectory(Directory);
						FTPShellCommon.FtpCurrentRemoteDir = FTPShellCommon.ClientFTP.GetWorkingDirectory();
						return true;
					}
					else
					{
						// Directory doesn't exist, go to the old directory
						throw new Kernel.Exceptions.FTPFilesystemException(Translate.DoTranslation("Directory {0} not found."), Directory);
					}
				}
				else
				{
					throw new ArgumentNullException(Directory, Translate.DoTranslation("Enter a remote directory. \"..\" to go back"));
				}
			}
			else
			{
				throw new InvalidOperationException(Translate.DoTranslation("You must connect to a server before changing directory"));
			}
			return false;
		}

		public static bool FTPChangeLocalDir(string Directory)
		{
			if (!string.IsNullOrEmpty(Directory))
			{
				string targetDir;
				targetDir = $"{FTPShellCommon.FtpCurrentDirectory}/{Directory}";
				Files.Filesystem.ThrowOnInvalidPath(targetDir);

				// Check if folder exists
				if (Checking.FolderExists(targetDir))
				{
					// Parse written directory
					var parser = new DirectoryInfo(targetDir);
					FTPShellCommon.FtpCurrentDirectory = parser.FullName;
					return true;
				}
				else
				{
					throw new Kernel.Exceptions.FTPFilesystemException(Translate.DoTranslation("Local directory {0} doesn't exist."), Directory);
				}
			}
			else
			{
				throw new ArgumentNullException(Directory, Translate.DoTranslation("Enter a local directory. \"..\" to go back."));
			}
			return false;
		}

		/// <summary>
		/// Move file or directory to another area, or rename the file
		/// </summary>
		/// <param name="Source">Source file or folder</param>
		/// <param name="Target">Target file or folder</param>
		/// <returns>True if successful; False if unsuccessful</returns>
		/// <exception cref="InvalidOperationException"></exception>
		public static bool FTPMoveItem(string Source, string Target)
		{
			if (FTPShellCommon.FtpConnected)
			{
				var Success = default(bool);

				// Begin the moving process
				string SourceFile = Source.Split('/').Last();
				DebugWriter.Wdbg(DebugLevel.I, "Moving from {0} to {1} with the source file of {2}...", Source, Target, SourceFile);
				if (FTPShellCommon.ClientFTP.DirectoryExists(Source))
				{
					Success = FTPShellCommon.ClientFTP.MoveDirectory(Source, Target);
				}
				else if (FTPShellCommon.ClientFTP.FileExists(Source) & FTPShellCommon.ClientFTP.DirectoryExists(Target))
				{
					Success = FTPShellCommon.ClientFTP.MoveFile(Source, Target + SourceFile);
				}
				else if (FTPShellCommon.ClientFTP.FileExists(Source))
				{
					Success = FTPShellCommon.ClientFTP.MoveFile(Source, Target);
				}
				DebugWriter.Wdbg(DebugLevel.I, "Moved. Result: {0}", Success);
				return Success;
			}
			else
			{
				throw new InvalidOperationException(Translate.DoTranslation("You must connect to server before performing transmission."));
			}
			return false;
		}

		/// <summary>
		/// Copy file or directory to another area, or rename the file
		/// </summary>
		/// <param name="Source">Source file or folder</param>
		/// <param name="Target">Target file or folder</param>
		/// <returns>True if successful; False if unsuccessful</returns>
		/// <exception cref="InvalidOperationException"></exception>
		public static bool FTPCopyItem(string Source, string Target)
		{
			if (FTPShellCommon.FtpConnected)
			{
				bool Success = true;
				var Result = default(object);

				// Begin the copying process
				string SourceFile = Source.Split('/').Last();
				DebugWriter.Wdbg(DebugLevel.I, "Copying from {0} to {1} with the source file of {2}...", Source, Target, SourceFile);
				if (FTPShellCommon.ClientFTP.DirectoryExists(Source))
				{
					FTPShellCommon.ClientFTP.DownloadDirectory(Paths.TempPath + "/FTPTransfer", Source);
					Result = FTPShellCommon.ClientFTP.UploadDirectory(Paths.TempPath + "/FTPTransfer/" + Source, Target);
				}
				else if (FTPShellCommon.ClientFTP.FileExists(Source) & FTPShellCommon.ClientFTP.DirectoryExists(Target))
				{
					FTPShellCommon.ClientFTP.DownloadFile(Paths.TempPath + "/FTPTransfer/" + SourceFile, Source);
					Result = FTPShellCommon.ClientFTP.UploadFile(Paths.TempPath + "/FTPTransfer/" + SourceFile, Target + "/" + SourceFile);
				}
				else if (FTPShellCommon.ClientFTP.FileExists(Source))
				{
					FTPShellCommon.ClientFTP.DownloadFile(Paths.TempPath + "/FTPTransfer/" + SourceFile, Source);
					Result = FTPShellCommon.ClientFTP.UploadFile(Paths.TempPath + "/FTPTransfer/" + SourceFile, Target);
				}
				Directory.Delete(Paths.TempPath + "/FTPTransfer", true);

				// See if copied successfully
				/* TODO ERROR: Skipped WarningDirectiveTrivia
				#Disable Warning BC42104
				*/
				if (Result.GetType() == typeof(List<FtpResult>))
				{
					foreach (FtpResult FileResult in (IEnumerable)Result)
					{
						if (FileResult.IsFailed)
						{
							DebugWriter.Wdbg(DebugLevel.E, "Transfer for {0} failed: {1}", FileResult.Name, FileResult.Exception.Message);
							DebugWriter.WStkTrc(FileResult.Exception);
							Success = false;
						}
					}
				}
				else if (Result.GetType() == typeof(FtpStatus))
				{
					if (((FtpStatus)Convert.ToInt32(Result)).IsFailure())
					{
						DebugWriter.Wdbg(DebugLevel.E, "Transfer failed");
						Success = false;
					}
				}
				/* TODO ERROR: Skipped WarningDirectiveTrivia
				#Enable Warning BC42104
				*/
				DebugWriter.Wdbg(DebugLevel.I, "Copied. Result: {0}", Success);
				return Success;
			}
			else
			{
				throw new InvalidOperationException(Translate.DoTranslation("You must connect to server before performing transmission."));
			}
			return false;
		}

		/// <summary>
		/// Changes the permissions of a remote file
		/// </summary>
		/// <param name="Target">Target file</param>
		/// <param name="Chmod">Permissions in CHMOD format. See https://man7.org/linux/man-pages/man2/chmod.2.html chmod(2) for more info.</param>
		/// <returns>True if successful; False if unsuccessful</returns>
		public static bool FTPChangePermissions(string Target, int Chmod)
		{
			if (FTPShellCommon.FtpConnected)
			{
				try
				{
					FTPShellCommon.ClientFTP.Chmod(Target, Chmod);
					return true;
				}
				catch (Exception ex)
				{
					DebugWriter.Wdbg(DebugLevel.E, "Error setting permissions ({0}) to file {1}: {2}", Chmod, Target, ex.Message);
					DebugWriter.WStkTrc(ex);
				}
			}
			else
			{
				throw new InvalidOperationException(Translate.DoTranslation("You must connect to server before performing filesystem operations."));
			}
			return false;
		}

	}
}
